pkg/fake_gcloud/lib/mem_datastore.dart

```diff
@@ -34,14 +34,34 @@ class MemDatastore implements Datastore {
       List<Entity> autoIdInserts,
       List<Key> deletes,
       Transaction transaction}) async {
-    deletes?.forEach((key) => _entities.remove(key));
-    inserts?.forEach((e) {
-      _entities[e.key] = e;
-    });
+    inserts ??= <Entity>[];
+    deletes ??= <Key>[];
+
     if (autoIdInserts != null && autoIdInserts.isNotEmpty) {
       throw UnimplementedError(
           'fake_gcloud.Datastore.autoIdInserts is not implemented.');
     }
+
+    // https://cloud.google.com/datastore/docs/concepts/transactions#what_can_be_done_in_a_transaction
+    if (inserts.length + deletes.length > 500) {
+      throw DatastoreError('Too many entities in the transaction.');
+    }
+
+    final newKeys = inserts.map((i) => i.key).toSet();
+
+    // check if updated key is deleted
+    for (final deletedKey in deletes) {
+      if (newKeys.contains(deletedKey)) {
+        throw DatastoreError('Conflicting update and delete on $deletedKey');
+      }
+    }
+
+    // execute commit
+    deletes.forEach((key) => _entities.remove(key));
+    inserts.forEach((e) {
+      _entities[e.key] = e;
+    });
```

Comment:
Can we documentation the limitation of these checks... From what I can see this doesn't ensure serializability, which is the important part of here.

We could track the keys that have been mutated since the `Transaction` was created to ensure that there is no conflicts. Or we could simply reject any transaction that overlaps with another transaction :D

Reply:
good points! I'll add TODOs for that.